File: internal/schema/schema.go
// Package schema parses a template's field schema and validates document data
// against it. See SPEC.md §6.
package schema

import (
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"strings"
	"time"
)

// Types accepted in a schema.
var Types = []string{"text", "textarea", "number", "money", "date", "bool", "select", "object", "list", "asset", "sequence"}

var keyRe = regexp.MustCompile(`^[a-z][a-z0-9_]*$`)
var dateRe = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}$`)

// Option of a select field.
type Option struct {
	Value string `json:"value"`
	Label string `json:"label,omitempty"`
}

// Field definition. Pointer fields distinguish "unset" from zero.
type Field struct {
	Key      string `json:"key"`
	Type     string `json:"type"`
	Label    string `json:"label,omitempty"`
	Help     string `json:"help,omitempty"`
	Required *bool  `json:"required,omitempty"`
	Default  any    `json:"default,omitempty"`

	Placeholder string   `json:"placeholder,omitempty"`
	Pattern     string   `json:"pattern,omitempty"`
	Min         *float64 `json:"min,omitempty"`
	Max         *float64 `json:"max,omitempty"`
	Step        *float64 `json:"step,omitempty"`
	Currency    string   `json:"currency,omitempty"`
	Options     []Option `json:"options,omitempty"`
	Fields      []Field  `json:"fields,omitempty"`
	Fragment    string   `json:"fragment,omitempty"`
	Accept      string   `json:"accept,omitempty"`
	Format      string   `json:"format,omitempty"`
	Reset       string   `json:"reset,omitempty"`
}

// IsRequired reports the effective required flag (default true).
func (f Field) IsRequired() bool { return f.Required == nil || *f.Required }

// DisplayLabel is the label or, when empty, the key.
func (f Field) DisplayLabel() string {
	if f.Label != "" {
		return f.Label
	}
	return f.Key
}

// Schema is the root: a flat list of fields; object and list nest.
type Schema struct {
	Fields []Field `json:"fields"`
}

// Error is a validation problem at a data path such as "items.2.amount".
type Error struct {
	Path    string `json:"path"`
	Message string `json:"message"`
}

func (e Error) Error() string { return e.Path + ": " + e.Message }

// Errors is a list of validation errors that also satisfies error.
type Errors []Error

func (es Errors) Error() string {
	parts := make([]string, len(es))
	for i, e := range es {
		parts[i] = e.Error()
	}
	return strings.Join(parts, "; ")
}

// Parse decodes and checks a schema. An empty input yields an empty schema.
func Parse(raw []byte) (*Schema, error) {
	s := &Schema{}
	if len(strings.TrimSpace(string(raw))) == 0 || string(raw) == "null" {
		return s, nil
	}
	if err := json.Unmarshal(raw, s); err != nil {
		return nil, fmt.Errorf("schema: %w", err)
	}
	if err := checkFields(s.Fields, "", 0); err != nil {
		return nil, err
	}
	return s, nil
}

func checkFields(fields []Field, path string, depth int) error {
	seen := map[string]bool{}
	for _, f := range fields {
		p := f.Key
		if path != "" {
			p = path + "." + f.Key
		}
		if !keyRe.MatchString(f.Key) {
			return fmt.Errorf("schema: invalid key %q at %q (want [a-z][a-z0-9_]*)", f.Key, path)
		}
		if seen[f.Key] {
			return fmt.Errorf("schema: duplicate key %q", p)
		}
		seen[f.Key] = true
		if !contains(Types, f.Type) {
			return fmt.Errorf("schema: field %q has unknown type %q", p, f.Type)
		}
		switch f.Type {
		case "object", "list":
			if len(f.Fields) == 0 {
				return fmt.Errorf("schema: field %q (%s) needs sub-fields", p, f.Type)
			}
			if depth >= 2 {
				return fmt.Errorf("schema: field %q nests too deep (max 3 levels)", p)
			}
			if err := checkFields(f.Fields, p, depth+1); err != nil {
				return err
			}
		case "select":
			if len(f.Options) == 0 {
				return fmt.Errorf("schema: select %q needs options", p)
			}
		case "sequence":
			if depth > 0 {
				return fmt.Errorf("schema: sequence %q must be top-level", p)
			}
			if f.Reset != "" && f.Reset != "never" && f.Reset != "year" {
				return fmt.Errorf("schema: sequence %q: reset must be \"never\" or \"year\"", p)
			}
		case "text":
			if f.Pattern != "" {
				if _, err := regexp.Compile(f.Pattern); err != nil {
					return fmt.Errorf("schema: field %q: bad pattern: %v", p, err)
				}
			}
		}
	}
	return nil
}

// Validate checks data against the schema. With strict=false only shape and
// types are checked (drafts may be incomplete); with strict=true required
// fields and list minimums are enforced too (render time).
func (s *Schema) Validate(data map[string]any, strict bool) Errors {
	var errs Errors
	validateFields(s.Fields, data, "", strict, &errs)
	return errs
}

func validateFields(fields []Field, data map[string]any, path string, strict bool, errs *Errors) {
	known := map[string]Field{}
	for _, f := range fields {
		known[f.Key] = f
	}
	keys := make([]string, 0, len(data))
	for k := range data {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		if _, ok := known[k]; !ok {
			*errs = append(*errs, Error{join(path, k), "unknown field"})
		}
	}
	for _, f := range fields {
		p := join(path, f.Key)
		v, present := data[f.Key]
		if !present || isEmpty(v) {
			if strict && f.IsRequired() {
				*errs = append(*errs, Error{p, "required"})
			}
			continue
		}
		switch f.Type {
		case "text", "textarea", "asset", "sequence", "select":
			str, ok := v.(string)
			if !ok {
				*errs = append(*errs, Error{p, "must be a string"})
				continue
			}
			if f.Type == "text" && f.Pattern != "" {
				if re, err := regexp.Compile(f.Pattern); err == nil && !re.MatchString(str) {
					*errs = append(*errs, Error{p, "does not match pattern"})
				}
			}
			if f.Type == "select" {
				found := false
				for _, o := range f.Options {
					if o.Value == str {
						found = true
					}
				}
				if !found {
					*errs = append(*errs, Error{p, "not one of the options"})
				}
			}
		case "number", "money":
			n, ok := toFloat(v)
			if !ok {
				*errs = append(*errs, Error{p, "must be a number"})
				continue
			}
			if f.Min != nil && n < *f.Min {
				*errs = append(*errs, Error{p, fmt.Sprintf("must be ≥ %v", *f.Min)})
			}
			if f.Max != nil && n > *f.Max {
				*errs = append(*errs, Error{p, fmt.Sprintf("must be ≤ %v", *f.Max)})
			}
		case "date":
			str, ok := v.(string)
			if !ok || !dateRe.MatchString(str) {
				*errs = append(*errs, Error{p, "must be a date (YYYY-MM-DD)"})
				continue
			}
			if _, err := time.Parse("2006-01-02", str); err != nil {
				*errs = append(*errs, Error{p, "invalid date"})
			}
		case "bool":
			if _, ok := v.(bool); !ok {
				*errs = append(*errs, Error{p, "must be true or false"})
			}
		case "object":
			m, ok := v.(map[string]any)
			if !ok {
				*errs = append(*errs, Error{p, "must be an object"})
				continue
			}
			validateFields(f.Fields, m, p, strict, errs)
		case "list":
			arr, ok := v.([]any)
			if !ok {
				*errs = append(*errs, Error{p, "must be a list"})
				continue
			}
			if strict && f.Min != nil && float64(len(arr)) < *f.Min {
				*errs = append(*errs, Error{p, fmt.Sprintf("needs at least %v rows", *f.Min)})
			}
			if f.Max != nil && float64(len(arr)) > *f.Max {
				*errs = append(*errs, Error{p, fmt.Sprintf("at most %v rows", *f.Max)})
			}
			for i, item := range arr {
				m, ok := item.(map[string]any)
				if !ok {
					*errs = append(*errs, Error{fmt.Sprintf("%s.%d", p, i), "must be an object"})
					continue
				}
				validateFields(f.Fields, m, fmt.Sprintf("%s.%d", p, i), strict, errs)
			}
		}
	}
}

// isEmpty treats nil, "" and empty lists as "not provided". Note that a list
// with zero rows is empty (so `min` applies), but false and 0 are values.
func isEmpty(v any) bool {
	switch x := v.(type) {
	case nil:
		return true
	case string:
		return x == ""
	case []any:
		return len(x) == 0
	}
	return false
}

// Empty builds a data object with every field present (empty strings, zero
// lists, nested objects) so that forms and templates have a stable shape.
// Defaults (including "today" for dates) are applied.
func (s *Schema) Empty(now time.Time) map[string]any {
	return emptyFields(s.Fields, now)
}

func emptyFields(fields []Field, now time.Time) map[string]any {
	m := map[string]any{}
	for _, f := range fields {
		m[f.Key] = f.EmptyValue(now)
	}
	return m
}

// EmptyValue is the initial value of a single field.
func (f Field) EmptyValue(now time.Time) any {
	switch f.Type {
	case "object":
		return emptyFields(f.Fields, now)
	case "list":
		rows := []any{}
		if f.Min != nil {
			for i := 0; i < int(*f.Min); i++ {
				rows = append(rows, emptyFields(f.Fields, now))
			}
		}
		return rows
	case "date":
		if d, ok := f.Default.(string); ok {
			if d == "today" {
				return now.Format("2006-01-02")
			}
			return d
		}
		return ""
	case "bool":
		if b, ok := f.Default.(bool); ok {
			return b
		}
		return false
	case "number", "money":
		if n, ok := toFloat(f.Default); ok {
			return n
		}
		return nil
	default:
		if d, ok := f.Default.(string); ok {
			return d
		}
		return ""
	}
}

// ApplyDefaults fills missing/empty top-level keys with their empty value
// (including "today" dates) and returns the same map for convenience.
func (s *Schema) ApplyDefaults(data map[string]any, now time.Time) map[string]any {
	if data == nil {
		data = map[string]any{}
	}
	for _, f := range s.Fields {
		if v, ok := data[f.Key]; !ok || isEmpty(v) {
			if f.Type == "sequence" {
				continue // assigned by the sequence hook
			}
			data[f.Key] = f.EmptyValue(now)
		}
	}
	return data
}

// ResetForDuplicate prepares data for "duplicate": sequence values are
// cleared so they get reassigned and dates with default "today" are reset.
func (s *Schema) ResetForDuplicate(data map[string]any, now time.Time) map[string]any {
	for _, f := range s.Fields {
		switch f.Type {
		case "sequence":
			data[f.Key] = ""
		case "date":
			if d, _ := f.Default.(string); d == "today" {
				data[f.Key] = now.Format("2006-01-02")
			}
		}
	}
	return data
}

// SequenceFields returns the top-level sequence fields.
func (s *Schema) SequenceFields() []Field {
	var out []Field
	for _, f := range s.Fields {
		if f.Type == "sequence" {
			out = append(out, f)
		}
	}
	return out
}

// SequencePeriod is the counter bucket for a sequence field: the year for
// reset "year", "all" otherwise.
func SequencePeriod(f Field, now time.Time) string {
	if f.Reset == "year" {
		return now.Format("2006")
	}
	return "all"
}

var seqTokenRe = regexp.MustCompile(`\{(n(?::\d+)?|yy|yyyy|mm)\}`)

// FormatSequence renders a sequence format such as "HN-{yy}-{n:3}".
// Tokens: {n}, {n:3} (zero-padded), {yy}, {yyyy}, {mm}. An empty format means "{n}".
func FormatSequence(format string, n int, now time.Time) string {
	if format == "" {
		format = "{n}"
	}
	return seqTokenRe.ReplaceAllStringFunc(format, func(tok string) string {
		switch inner := tok[1 : len(tok)-1]; {
		case inner == "yy":
			return now.Format("06")
		case inner == "yyyy":
			return now.Format("2006")
		case inner == "mm":
			return now.Format("01")
		case strings.HasPrefix(inner, "n:"):
			var w int
			fmt.Sscanf(inner[2:], "%d", &w)
			return fmt.Sprintf("%0*d", w, n)
		default:
			return fmt.Sprintf("%d", n)
		}
	})
}

// FindByKind returns the sub-schema of the first object field tagged with the
// given fragment kind (used by the fragments page to build a form).
func (s *Schema) FindByKind(kind string) *Field {
	for i := range s.Fields {
		if s.Fields[i].Type == "object" && s.Fields[i].Fragment == kind {
			return &s.Fields[i]
		}
	}
	return nil
}

func join(path, key string) string {
	if path == "" {
		return key
	}
	return path + "." + key
}

func toFloat(v any) (float64, bool) {
	switch x := v.(type) {
	case float64:
		return x, true
	case float32:
		return float64(x), true
	case int:
		return float64(x), true
	case int64:
		return float64(x), true
	case json.Number:
		f, err := x.Float64()
		return f, err == nil
	}
	return 0, false
}

func contains(list []string, s string) bool {
	for _, x := range list {
		if x == s {
			return true
		}
	}
	return false
}

File: ui/src/lib/schema.ts
/** Client-side mirror of internal/schema (SPEC.md §6): types, empty values, validation. */

export const TYPES = ['text', 'textarea', 'number', 'money', 'date', 'bool', 'select', 'object', 'list', 'asset', 'sequence'] as const
export type FieldType = (typeof TYPES)[number]

export interface Option { value: string; label?: string }

export interface Field {
  key: string
  type: FieldType
  label?: string
  help?: string
  required?: boolean
  default?: any
  placeholder?: string
  pattern?: string
  min?: number
  max?: number
  step?: number
  currency?: string
  options?: Option[]
  fields?: Field[]
  fragment?: string
  accept?: string
  format?: string
  reset?: 'never' | 'year'
}

export interface Schema { fields: Field[] }
export interface Err { path: string; message: string }

const keyRe = /^[a-z][a-z0-9_]*$/
const dateRe = /^\d{4}-\d{2}-\d{2}$/

export const label = (f: Field) => f.label || f.key
export const required = (f: Field) => f.required !== false
export const today = () => new Date().toISOString().slice(0, 10)

/** Parse schema JSON text; returns the schema or an error message. */
export function parseSchema(text: string): { schema?: Schema; error?: string } {
  if (!text.trim()) return { schema: { fields: [] } }
  let raw: any
  try { raw = JSON.parse(text) } catch (e: any) { return { error: 'Invalid JSON: ' + e.message } }
  return checkSchema(raw)
}

export function checkSchema(raw: any): { schema?: Schema; error?: string } {
  if (!raw || typeof raw !== 'object' || !Array.isArray(raw.fields)) return { error: 'Schema must be an object with a "fields" array' }
  const err = checkFields(raw.fields, '', 0)
  return err ? { error: err } : { schema: raw as Schema }
}

function checkFields(fields: any[], path: string, depth: number): string {
  const seen = new Set<string>()
  for (const f of fields) {
    const p = path ? `${path}.${f?.key}` : f?.key
    if (!f || typeof f !== 'object') return `field at "${path || 'root'}" must be an object`
    if (typeof f.key !== 'string' || !keyRe.test(f.key)) return `invalid key "${f.key}" at "${path || 'root'}" (want [a-z][a-z0-9_]*)`
    if (seen.has(f.key)) return `duplicate key "${p}"`
    seen.add(f.key)
    if (!TYPES.includes(f.type)) return `field "${p}" has unknown type "${f.type}"`
    if (f.type === 'object' || f.type === 'list') {
      if (!Array.isArray(f.fields) || !f.fields.length) return `field "${p}" (${f.type}) needs sub-fields`
      if (depth >= 2) return `field "${p}" nests too deep (max 3 levels)`
      const e = checkFields(f.fields, p, depth + 1)
      if (e) return e
    }
    if (f.type === 'select' && (!Array.isArray(f.options) || !f.options.length)) return `select "${p}" needs options`
    if (f.type === 'sequence') {
      if (depth > 0) return `sequence "${p}" must be top-level`
      if (f.reset && f.reset !== 'never' && f.reset !== 'year') return `sequence "${p}": reset must be "never" or "year"`
    }
    if (f.type === 'text' && f.pattern) {
      try { new RegExp(f.pattern) } catch (e: any) { return `field "${p}": bad pattern: ${e.message}` }
    }
  }
  return ''
}

/** Initial value of a field (defaults applied, "today" for dates). */
export function emptyValue(f: Field): any {
  switch (f.type) {
    case 'object': return emptyData(f.fields ?? [])
    case 'list': return Array.from({ length: f.min ?? 0 }, () => emptyData(f.fields ?? []))
    case 'date': return f.default === 'today' ? today() : (typeof f.default === 'string' ? f.default : '')
    case 'bool': return typeof f.default === 'boolean' ? f.default : false
    case 'number': case 'money': return typeof f.default === 'number' ? f.default : null
    default: return typeof f.default === 'string' ? f.default : ''
  }
}

export const emptyData = (fields: Field[]): Record<string, any> =>
  Object.fromEntries(fields.map(f => [f.key, emptyValue(f)]))

/** Fill missing keys so the form has a stable shape (keeps existing values). */
export function normalize(fields: Field[], data: any): Record<string, any> {
  const out: Record<string, any> = data && typeof data === 'object' && !Array.isArray(data) ? data : {}
  for (const f of fields) {
    const v = out[f.key]
    if (f.type === 'object') out[f.key] = normalize(f.fields ?? [], v)
    else if (f.type === 'list') out[f.key] = Array.isArray(v) ? v.map(row => normalize(f.fields ?? [], row)) : emptyValue(f)
    else if (v === undefined || v === null) out[f.key] = emptyValue(f)
  }
  return out
}

const isEmpty = (v: any) => v === null || v === undefined || v === '' || (Array.isArray(v) && v.length === 0)

/** Validate data; strict enforces required + list min (render time). */
export function validate(fields: Field[], data: any, strict: boolean, path = '', errs: Err[] = []): Err[] {
  const known = new Set(fields.map(f => f.key))
  for (const k of Object.keys(data ?? {})) if (!known.has(k)) errs.push({ path: join(path, k), message: 'unknown field' })
  for (const f of fields) {
    const p = join(path, f.key)
    const v = data?.[f.key]
    if (isEmpty(v)) { if (strict && required(f)) errs.push({ path: p, message: 'required' }); continue }
    switch (f.type) {
      case 'text': case 'textarea': case 'asset': case 'sequence': case 'select':
        if (typeof v !== 'string') { errs.push({ path: p, message: 'must be a string' }); break }
        if (f.type === 'text' && f.pattern && !new RegExp(f.pattern).test(v)) errs.push({ path: p, message: 'does not match pattern' })
        if (f.type === 'select' && !f.options?.some(o => o.value === v)) errs.push({ path: p, message: 'not one of the options' })
        break
      case 'number': case 'money':
        if (typeof v !== 'number' || Number.isNaN(v)) { errs.push({ path: p, message: 'must be a number' }); break }
        if (f.min !== undefined && v < f.min) errs.push({ path: p, message: `must be ≥ ${f.min}` })
        if (f.max !== undefined && v > f.max) errs.push({ path: p, message: `must be ≤ ${f.max}` })
        break
      case 'date':
        if (typeof v !== 'string' || !dateRe.test(v) || Number.isNaN(Date.parse(v))) errs.push({ path: p, message: 'must be a date (YYYY-MM-DD)' })
        break
      case 'bool':
        if (typeof v !== 'boolean') errs.push({ path: p, message: 'must be true or false' })
        break
      case 'object':
        if (!v || typeof v !== 'object' || Array.isArray(v)) { errs.push({ path: p, message: 'must be an object' }); break }
        validate(f.fields ?? [], v, strict, p, errs)
        break
      case 'list':
        if (!Array.isArray(v)) { errs.push({ path: p, message: 'must be a list' }); break }
        if (strict && f.min !== undefined && v.length < f.min) errs.push({ path: p, message: `needs at least ${f.min} rows` })
        if (f.max !== undefined && v.length > f.max) errs.push({ path: p, message: `at most ${f.max} rows` })
        v.forEach((row, i) => validate(f.fields ?? [], row, strict, `${p}.${i}`, errs))
        break
    }
  }
  return errs
}

/** Keep only the keys the schema declares (recursively) — drops leftovers from fragments or older schemas. */
export function pick(fields: Field[], data: any): Record<string, any> {
  const out: Record<string, any> = {}
  if (!data || typeof data !== 'object') return out
  for (const f of fields) {
    const v = data[f.key]
    if (v === undefined) continue
    if (f.type === 'object') out[f.key] = pick(f.fields ?? [], v)
    else if (f.type === 'list') out[f.key] = Array.isArray(v) ? v.map(row => pick(f.fields ?? [], row)) : v
    else out[f.key] = v
  }
  return out
}

/** Deep clone that also works on Svelte state proxies. */
export const clone = <T>(v: T): T => JSON.parse(JSON.stringify(v ?? null)) ?? ({} as T)

const join = (a: string, b: string) => (a ? `${a}.${b}` : b)

/** Object fields tagged with a fragment kind, at any depth. */
export function fragmentKinds(fields: Field[]): string[] {
  const out = new Set<string>()
  const walk = (fs: Field[]) => fs.forEach(f => { if (f.type === 'object' && f.fragment) out.add(f.fragment); if (f.fields) walk(f.fields) })
  walk(fields)
  return [...out]
}

/** First object field with the given fragment kind. */
export function findByKind(fields: Field[], kind: string): Field | undefined {
  for (const f of fields) {
    if (f.type === 'object' && f.fragment === kind) return f
    if (f.fields) { const r = findByKind(f.fields, kind); if (r) return r }
  }
  return undefined
}
